notebook/exploration-reports/2022.03-ipld-url-scheme.md

```diff
@@ -0,0 +1,106 @@
+With that in mind, there should be a clear mapping between HTTP-like APIs via gateways and URL handlers, and CLI arguments.
+In order to make it easier to come to an initial spec, we will focus on a minimal set of features and make subsequent specs with support for features like ADLs, Schemas, and advanced selectors.
+
+## Initial Functionality
+
+The initial functionality focuses on being able to GET and PUT IPLD data to and from a URL.
+For now, this will be done by encoding individual blocks of IPLD data using codecs with a default to the [DAG-JSON](https://ipld.io/specs/codecs/dag-json/spec/) codec.
+Some initial support for pathing is also included for convenience.
+
+### GET `ipld://{CID}/{Path} Headers {Accept="application/vnd.ipld.dag-json"} => Data`
+
+The most basic mapping is to get some data from a CID.
+
+The `CID` parameter in the URL MUST be in the CIDv1 format or any other format that is not case sensitive.
+This is necessary since the URL scheme considers the `hostname` portion to be case-insensitive and many URL parsers (such as the one in Chrome) will automatically convert it to lowercase. 
+
+`Data` is the response body which contaisn the encoded IPLD data for the given CID+Path.
+Note that any `Link`s within the IPLD data will not be resolved, and you'll need to conver them to URLs and fetch additional data yourself, or add path segments to the URL to traverse into them.
+
+The `Accept` header may be optionally provided to specify the codec that should be used in the response.
+This follows the same conventions as in the [IPFS gateway](https://github.com/ipfs/go-ipfs/pull/8758).
+Other codecs that may be targetted are DAG-CBOR.
+TODO: `application/vnd.ipld.dag-json` or just `application/json`?
+
+The `Path` can be used to specify segments separated by a `/` to use with basic traversal,
+if it is supplied the IPLD Node at that location relative to the CID will be returned.
+If there path does not lead to a valid IPLD node, an error will be returned instead, e.g. `HTTP 404`.
+
+### PUT `ipld://{CID}/{path} Headers {Content-Type="application/vnd.ipld.dag-json"} Body={encoded IPLD data} => URL`
+
+The `{CID}` should be the "base" CID to extend data from.
+Sadly if we are to use URLs (TODO: why should we use URLs?), we must have some sort of CID in the hostname.
+TODO: Document some inline CIDs to use. IPFS has this empty directory CID `ipfs://bafyaabakaieac` which would be cool to have analogs for for empty `{}` object and `[]` array nodes.
+
+The `Body` of the request should contain the encoded IPLD data you wish to upload.
+
+The `Content-Type` may be optionally provided to specify the encoding of the `Body`.
+By default it will be assumed to be `DAG-JSON`.
+
+If a `Path` is specified, intermediate Nodes will be created and added to the Node specified by the CID.
```

Comment:
Does this strictly stick to the integer components of a path, where if it encounters a `0` it'll create an array for you? And what if the path has a `1` (or later) in it? Do we make an array that has that many elements that are ... empty? Because we'd have to know what "empty" is (back to the _unit_ discussion).

Reply:
I think to be safe it would be good to generate intermediate Map Nodes if a node doesn't already exist at that path otherwise it might be too magical. 😅 E.g. what if somebody is iterating through a list of names and one of them just happens to be an integer.

Also, should we error out if somebody tries to add a named key to a node which is an array?

Reply:
I think this could also be augmented down the line with Schemas so you can give better type hints for individual segments of the path.

Reply:
Yeah, I think erroring out when there isn't something obvious and having fancier behavior in the future would be best for this initial version. I'll add some notes about how lists should be handled in the meantime though.

Reply:
Thinking more on this, maybe we should error out when there's no intermediate node? E.g., users would need to manually add intermediate nodes with separate requests as they're building up the DAG.

Reply:
@rvagg Mind weighing in on whether we should error out when there's no intermediate node?

Reply:
Hmmmm, I don't have a strong opinion. There's a nice convenience to automatically handling intermediate nodes (`mkdir -p`, `rm -r` - but notice they require special flags!). It might be tedious to have to do them all, one-by-one—round-trips get expensive in content-addressed land. But on the other hand, does it indicate a possible error if you're encountering missing intermediates? Perhaps so.

Your call if you're able to form a firmer opinion.

Reply:
I've been chewing on it and I think the convenience outweighs the downsides. 😅

I've got auto-generation (-p) for ipfs/ipns/hyper in Agregore at the moment and not having it made a lot of things more annoying. Ultimately, I think developer UX is really important for this stuff and streamlining is usually worth it.